### ISSUE TITLE
Trouble understanding sharded gradient accumulate?

### ISSUE THREAD

**Issue by rohan-mehta-1024** [stanford-crfm/levanter#342]
I am having some trouble understanding how ```accumulate_gradients_sharded``` works. In particular, when it reshapes the batches intro microbatches via the function below. Aren't the batches already sharded over the devices due to the sharded data loader? Also, once we flatten the Batch dimension into (AccumStep, Microbatch) how do we shard it with the parameter axis mapping? The default parameter axis mapping only shards the embedding dimension across the specified physical axis, right?  And even if it did shard the Batch dimension, we've just exchanged Batch for (AccumStep, Microbatch), so it seems like it still wouldn't work...I'm not sure what I'm getting wrong here???

```def _reshape_for_microbatch(Batch: Axis, Microbatch: Axis, AccumStep: Axis, inputs, axis_mapping):
    def _reshape(x):
        if isinstance(x, hax.NamedArray):
            if not x.has_axis(Batch.name):
                return x
            x = x.unflatten_axis(Batch, (AccumStep, Microbatch))
            return hax.shard_with_axis_mapping(x, axis_mapping)
        elif isinstance(x, jnp.ndarray):
            x = x.reshape((AccumStep.size, Microbatch.size) + x.shape[1:])
            return with_sharding_constraint(x, PartitionSpec(None, ResourceAxis.DATA, *(None,) * (len(x.shape) - 2)))
        else:
            assert jnp.isscalar(x)
            return x

    return jax.tree_util.tree_map(_reshape, inputs, is_leaf=is_named_array)```

**Comment by dlwh** [comment]
this is definitely not obvious and I should document it better, and i can probably make improvements.

First, basically any time you do "shape manipulation" (reshape, un/flatten_axes, etc) it's a good idea to use `with_sharding_constraint`/`shard_with_axis_mapping` to hint to the compiler what axes you want to (still) be sharded.
Perhaps I should do that automatically in Haliax.

I think that as an old holdover the parameter_axis_mapping in Levanter actually does map batch to data, which I should probably fix.

Microbatch has the same name as Batch so  has the same sharding properties.

Does that make sense?

**Comment by rohan-mehta-1024** [comment]
Sorry for the late response – this was very helpful, thanks! I pretty much understand everything going on here now. I do have one other slightly tangential question, but it seems like it would be cleaner to ask here than open up a whole new issue?  So, for example, in the `trainer.py` script, the decorator `@named_jit` is used a few times, e.g.: 


```
@named_jit(in_axis_resources=self.parameter_axis_mapping, axis_resources=self.compute_axis_mapping)
def fn(model, *batch, **batch_kwargs):
    pass
```
       
As far as I understand it, this shards the args according to in_axis_resources, then does the computation within the decorated function according to axis_resources, and re-shards the function's outputs according to out_axis_resources. In this case, as you mentioned above, it seems like parameter_axis_mapping shards both params and data. But is it possible to shard different arguments according to different mappings, e.g., if my function takes in model parameters and data and I have separate mappings for each (or would it be better to just feed in a merged dict as in_axis_resources)? Also, what is the efficiency vs idiomatic code tradeoff for pre-sharding model parameters and optimizer states once, when initializing them, versus using `hax.named_jit` to re-shard them right before every train step? It seems to me like the code in `trainer.py` is doing the latter (versus, e.g., the tutorials, which pre-sharded both model and optimizer, I believe?), though I could be wrong.

Thanks!

**Comment by dlwh** [comment]
No worries.

Yes I would like to add support for multiple axis mappings, but I think better is probably to move away from `in_axis_resources`/`out_axis_resources` and towards using `shard_with_axis_mapping` before and after. I haven't done that yet, but I think it's a better way of doing things.

the model and opt_state are created fully sharded inside trainer. 

```

**Comment by dlwh** [comment]
@rohan-mehta-1024 it's been a bit, but I have a new implementation that I think is much clearer. WDYT? https://github.com/stanford-crfm/levanter/blob/main/src/levanter/grad_accum.py#L30

**Comment by dlwh** [comment]
closing, but lemme know if it's not clear and i'll reopen
